wgpu-native/src/device.rs

```diff
@@ -269,34 +271,40 @@ impl DestroyedResources<back::Backend> {
-        let mut buffer_guard = HUB.buffers.write();
-
         for buffer_id in self.ready_to_map.drain(..) {
-            let buffer = &mut buffer_guard[buffer_id];
-            let mut operation = None;
-            std::mem::swap(&mut operation, &mut buffer.pending_map_operation);
-            match operation {
-                Some(BufferMapOperation::Read(range, callback, userdata)) => {
-                    if let Ok(ptr) = unsafe { raw.map_memory(&buffer.memory, range.clone()) } {
-                        if !buffer.memory_properties.contains(hal::memory::Properties::COHERENT) {
-                            unsafe { raw.invalidate_mapped_memory_ranges(iter::once((&buffer.memory, range.clone()))).unwrap()  }; // TODO
+            let (operation, result, ptr) = {
+                let mut buffer_guard = HUB.buffers.write();
+                let buffer = &mut buffer_guard[buffer_id];
+                let mut operation = None;
+                std::mem::swap(&mut operation, &mut buffer.pending_map_operation);
+                match operation {
+                    Some(BufferMapOperation::Read(range, callback, userdata)) => {
+                        if let Ok(ptr) = unsafe { raw.map_memory(&buffer.memory, range.clone()) } {
+                            if !buffer.memory_properties.contains(hal::memory::Properties::COHERENT) {
+                                unsafe { raw.invalidate_mapped_memory_ranges(iter::once((&buffer.memory, range.clone()))).unwrap()  }; // TODO
+                            }
+                            (BufferMapOperation::Read(range, callback, userdata), BufferMapAsyncStatus::Success, Some(ptr))
+                        } else {
+                            (BufferMapOperation::Read(range, callback, userdata), BufferMapAsyncStatus::Error, None)
                         }
-                        callback(BufferMapAsyncStatus::Success, ptr, userdata);
-                    } else {
-                        callback(BufferMapAsyncStatus::Error, std::ptr::null(), userdata);
-                    }
-                },
-                Some(BufferMapOperation::Write(range, callback, userdata)) => {
-                    if let Ok(ptr) = unsafe { raw.map_memory(&buffer.memory, range.clone()) } {
-                        if !buffer.memory_properties.contains(hal::memory::Properties::COHERENT) {
-                            buffer.mapped_write_ranges.push(range);
+                    },
+                    Some(BufferMapOperation::Write(range, callback, userdata)) => {
+                        if let Ok(ptr) = unsafe { raw.map_memory(&buffer.memory, range.clone()) } {
+                            if !buffer.memory_properties.contains(hal::memory::Properties::COHERENT) {
+                                buffer.mapped_write_ranges.push(range.clone());
+                            }
+                            (BufferMapOperation::Write(range, callback, userdata), BufferMapAsyncStatus::Success, Some(ptr))
```

Comment:
if the operation is always unchanged, can we just do something like `operation.clone().unwrap()` once at the end, instead of repeating it in each if/else arm?